File: backend/internal/brand/lint.go
// Brand linter + pre-export gate (doc 18 FR-7/FR-8), porting @hc/brandkit
// lint.ts + gate.ts. A pure rule engine over the design file (map[string]any)
// and a resolved kit: off-brand color/font (when locked), low-contrast text,
// logo misuse, and margin spacing. The gate turns the violations into an
// export-blocking decision under lintPolicy "block". Reuses the color/kit
// extraction helpers in enforcement.go so the live lint and the persist gate
// agree. The editor runs the same rules client-side.
package brand

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"

	"hycanvas/backend/internal/color"
)

const (
	aaNormal        = 4.5 // WCAG AA normal-text contrast
	aspectTolerance = 0.02
)

// LintViolation is one finding (matches the SDK BrandLintViolation).
type LintViolation struct {
	ID       string         `json:"id"`
	Kind     string         `json:"kind"`
	Severity string         `json:"severity"`
	NodeID   string         `json:"nodeId"`
	PageID   string         `json:"pageId"`
	Message  string         `json:"message"`
	Fix      map[string]any `json:"fix,omitempty"`
}

// GateResult is the pre-export decision (matches the SDK BrandLintResult).
type GateResult struct {
	Policy     string          `json:"policy"`
	Blocked    bool            `json:"blocked"`
	Violations []LintViolation `json:"violations"`
}

type lintKit struct {
	palette  []rgb
	families map[string]bool
	swapFont string
	logoIDs  map[string]bool
	logoMin  map[string]float64
	controls BrandControls
}

func numAt(m map[string]any, k string) float64 {
	switch n := m[k].(type) {
	case float64:
		return n
	case int:
		return float64(n)
	}
	return 0
}

func strAt(v any) string { s, _ := v.(string); return s }

// parseLintKit builds the lint inputs from a resolved kit row, reusing the
// palette/family extraction shared with the locked-save enforcement.
func parseLintKit(row *BrandKitRow) lintKit {
	bk := view(*row)
	k := lintKit{
		palette:  kitColors(bk),
		families: kitFontFamilies(bk),
		controls: bk.Controls,
		logoIDs:  map[string]bool{},
		logoMin:  map[string]float64{},
	}

	var fonts []map[string]any
	_ = json.Unmarshal(bk.Fonts, &fonts)
	for _, f := range fonts {
		fam := strAt(f["fontFamily"])
		if fam == "" {
			continue
		}
		if k.swapFont == "" || f["role"] == "body" {
			k.swapFont = fam
		}
	}

	var logos []map[string]any
	_ = json.Unmarshal(bk.Logos, &logos)
	for _, l := range logos {
		variants := asObjMap(l["variants"])
		for _, id := range []string{strAt(l["assetId"]), strAt(variants["dark"]), strAt(variants["light"])} {
			if id != "" {
				k.logoIDs[id] = true
			}
		}
		if id := strAt(l["assetId"]); id != "" {
			if m := numAt(l, "minSizePx"); m > 0 {
				k.logoMin[id] = m
			}
		}
	}
	return k
}

func nearestDistance(c rgb, palette []rgb) (float64, rgb) {
	best := math.Inf(1)
	var bc rgb
	for _, p := range palette {
		if d := color.DeltaE(c.r, c.g, c.b, p.r, p.g, p.b); d < best {
			best, bc = d, p
		}
	}
	return best, bc
}

// walkTyped visits each node (recursing children + mask child + boolean
// operands) for the contrast/logo/spacing passes that need node type/geometry.
func walkTyped(nodes []any, visit func(n map[string]any)) {
	for _, raw := range nodes {
		n := asObjMap(raw)
		if n == nil {
			continue
		}
		visit(n)
		walkTyped(asArrAny(n["children"]), visit)
		if n["type"] == "mask" {
			if ch := asObjMap(n["child"]); ch != nil {
				visit(ch)
			}
		}
		if n["type"] == "boolean" {
			walkTyped(asArrAny(n["operands"]), visit)
		}
	}
}

func pageBackground(page map[string]any) (rgb, bool) {
	bg := asObjMap(page["background"])
	if bg == nil {
		return rgb{}, false
	}
	var cs []rgb
	eachFillColor(bg, &cs)
	if len(cs) == 0 {
		return rgb{}, false
	}
	return cs[0], true
}

func lintFile(file map[string]any, kit lintKit) []LintViolation {
	out := []LintViolation{}
	if kit.controls.LintPolicy == "off" {
		return out
	}
	pages := asArrAny(file["pages"])

	// Off-brand colors.
	if kit.controls.LockColors && len(kit.palette) > 0 {
		seen := map[string]bool{}
		for _, pr := range pages {
			page := asObjMap(pr)
			pageID := strAt(page["id"])
			var colors []colorAt
			var fonts []fontAt
			if bg, ok := pageBackground(page); ok {
				colors = append(colors, colorAt{bg, pageID})
			}
			walkBrand(asArrAny(page["children"]), &colors, &fonts)
			for _, u := range colors {
				dist, near := nearestDistance(u.c, kit.palette)
				if dist <= colorTolerance {
					continue
				}
				hex := hexOf(u.c)
				key := "color:" + hex + "@" + u.nodeID
				if seen[key] {
					continue
				}
				seen[key] = true
				out = append(out, LintViolation{
					ID: key, Kind: "off-brand-color", Severity: "warn", NodeID: u.nodeID, PageID: pageID,
					Message: fmt.Sprintf("Color %s is not in the brand palette.", hex),
					Fix:     map[string]any{"kind": "snap_color", "from": hex, "to": hexOf(near)},
				})
			}
		}
	}

	// Off-brand fonts.
	if kit.controls.LockFonts && len(kit.families) > 0 {
		seen := map[string]bool{}
		for _, pr := range pages {
			page := asObjMap(pr)
			pageID := strAt(page["id"])
			var colors []colorAt
			var fonts []fontAt
			walkBrand(asArrAny(page["children"]), &colors, &fonts)
			for _, f := range fonts {
				if kit.families[strings.ToLower(strings.TrimSpace(f.family))] {
					continue
				}
				key := "font:" + f.family + "@" + f.nodeID
				if seen[key] {
					continue
				}
				seen[key] = true
				v := LintViolation{
					ID: key, Kind: "off-brand-font", Severity: "warn", NodeID: f.nodeID, PageID: pageID,
					Message: fmt.Sprintf("Font %q is not an approved brand font.", f.family),
				}
				if kit.swapFont != "" && !strings.EqualFold(strings.TrimSpace(kit.swapFont), strings.TrimSpace(f.family)) {
					v.Fix = map[string]any{"kind": "swap_font", "from": f.family, "to": kit.swapFont}
				}
				out = append(out, v)
			}
		}
	}

	// Low-contrast text (advisory whenever a kit is present).
	for _, pr := range pages {
		page := asObjMap(pr)
		pageID := strAt(page["id"])
		bg := rgb{1, 1, 1}
		if c, ok := pageBackground(page); ok {
			bg = c
		}
		walkTyped(asArrAny(page["children"]), func(n map[string]any) {
			if n["type"] != "text" {
				return
			}
			fg, ok := firstTextColor(n)
			if !ok {
				return
			}
			ratio := color.ContrastRatio(fg.r, fg.g, fg.b, bg.r, bg.g, bg.b)
			if ratio >= aaNormal {
				return
			}
			out = append(out, LintViolation{
				ID: "contrast:" + strAt(n["id"]), Kind: "low-contrast", Severity: "warn",
				NodeID: strAt(n["id"]), PageID: pageID,
				Message: fmt.Sprintf("Text contrast %.1f:1 is below WCAG AA (4.5:1).", ratio),
			})
		})
	}

	// Logo misuse.
	if len(kit.logoIDs) > 0 {
		for _, pr := range pages {
			page := asObjMap(pr)
			pageID := strAt(page["id"])
			walkTyped(asArrAny(page["children"]), func(n map[string]any) {
				if n["type"] != "image" {
					return
				}
				assetID := strAt(asObjMap(n["source"])["assetId"])
				if assetID == "" || !kit.logoIDs[assetID] {
					return
				}
				nodeID := strAt(n["id"])
				t := asObjMap(n["transform"])
				sx, sy := math.Abs(scaleOr1(t, "scaleX")), math.Abs(scaleOr1(t, "scaleY"))
				if sx > 0 && sy > 0 && math.Abs(sx-sy)/math.Max(sx, sy) > aspectTolerance {
					out = append(out, LintViolation{
						ID: "logo-aspect:" + nodeID, Kind: "logo-misuse", Severity: "error", NodeID: nodeID, PageID: pageID,
						Message: "Logo is distorted (non-uniform scale). Scale it uniformly.",
						Fix:     map[string]any{"kind": "restore_logo", "reason": "aspect"},
					})
				}
				for _, f := range asArrAny(n["fills"]) {
					if asObjMap(f)["type"] == "solid" {
						out = append(out, LintViolation{
							ID: "logo-recolor:" + nodeID, Kind: "logo-misuse", Severity: "error", NodeID: nodeID, PageID: pageID,
							Message: "Logo appears recolored. Use an approved logo variant instead.",
						})
						break
					}
				}
				if min := kit.logoMin[assetID]; min > 0 {
					if w := numAt(asObjMap(n["size"]), "width") * sx; w > 0 && w < min {
						out = append(out, LintViolation{
							ID: "logo-size:" + nodeID, Kind: "logo-misuse", Severity: "warn", NodeID: nodeID, PageID: pageID,
							Message: fmt.Sprintf("Logo is below its minimum size (%.0fpx).", min),
							Fix:     map[string]any{"kind": "restore_logo", "reason": "min_size"},
						})
					}
				}
			})
		}
	}

	// Spacing (opt-in via meta.brandMargin).
	if margin := numAt(asObjMap(file["meta"]), "brandMargin"); margin > 0 {
		for _, pr := range pages {
			page := asObjMap(pr)
			pageID := strAt(page["id"])
			pw, ph := numAt(page, "width"), numAt(page, "height")
			walkTyped(asArrAny(page["children"]), func(n map[string]any) {
				t, sz := asObjMap(n["transform"]), asObjMap(n["size"])
				if t == nil || sz == nil {
					return
				}
				x, y, nw, nh := numAt(t, "x"), numAt(t, "y"), numAt(sz, "width"), numAt(sz, "height")
				if x < margin || y < margin || x+nw > pw-margin || y+nh > ph-margin {
					out = append(out, LintViolation{
						ID: "spacing:" + strAt(n["id"]), Kind: "spacing", Severity: "info",
						NodeID: strAt(n["id"]), PageID: pageID,
						Message: fmt.Sprintf("Element extends into the %.0fpx brand margin.", margin),
					})
				}
			})
		}
	}

	return out
}

// firstTextColor returns a text node's dominant run color (style.color, then a
// solid run fill).
func firstTextColor(n map[string]any) (rgb, bool) {
	for _, para := range asArrAny(n["content"]) {
		for _, run := range asArrAny(asObjMap(para)["runs"]) {
			style := asObjMap(asObjMap(run)["style"])
			if c, ok := srgbOf(asObjMap(style["color"])); ok {
				return c, true
			}
			if sf := asObjMap(style["fill"]); sf != nil {
				var cs []rgb
				eachFillColor(sf, &cs)
				if len(cs) > 0 {
					return cs[0], true
				}
			}
		}
	}
	return rgb{}, false
}

func scaleOr1(m map[string]any, k string) float64 {
	if m == nil {
		return 1
	}
	if v, ok := m[k].(float64); ok {
		return v
	}
	return 1
}

// LintDesign returns the brand violations for a design (FR-7). Membership-gated.
// With no resolved kit it returns no violations.
func (s *Service) LintDesign(ctx context.Context, designID, userID string) ([]LintViolation, error) {
	ws, err := s.store.GetWorkspaceID(ctx, designID)
	if err != nil {
		return nil, err
	}
	if err := s.assertMember(ctx, userID, ws); err != nil {
		return nil, err
	}
	row, err := s.resolveKitRow(ctx, designID, ws)
	if err != nil {
		return nil, err
	}
	if row == nil {
		return []LintViolation{}, nil
	}
	file, err := s.store.FileFor(ctx, designID, ws)
	if err != nil {
		return nil, err
	}
	return lintFile(file, parseLintKit(row)), nil
}

// LintGate is the pre-export/publish brand gate (FR-8). With no kit or policy
// "off" it never blocks; under "block" any non-info violation blocks.
func (s *Service) LintGate(ctx context.Context, designID, userID string) (GateResult, error) {
	ws, err := s.store.GetWorkspaceID(ctx, designID)
	if err != nil {
		return GateResult{}, err
	}
	if err := s.assertMember(ctx, userID, ws); err != nil {
		return GateResult{}, err
	}
	row, err := s.resolveKitRow(ctx, designID, ws)
	if err != nil {
		return GateResult{}, err
	}
	if row == nil {
		return GateResult{Policy: "off", Blocked: false, Violations: []LintViolation{}}, nil
	}
	file, err := s.store.FileFor(ctx, designID, ws)
	if err != nil {
		return GateResult{}, err
	}
	kit := parseLintKit(row)
	violations := lintFile(file, kit)
	policy := kit.controls.LintPolicy
	blocked := false
	if policy == "block" {
		for _, v := range violations {
			if v.Severity != "info" {
				blocked = true
				break
			}
		}
	}
	return GateResult{Policy: policy, Blocked: blocked, Violations: violations}, nil
}

File: packages/brandkit/src/lint.ts
// The brand linter. A pure rule engine over a DesignFile + a
// LintBrandKit: it walks every page/node, then emits violations of five kinds
// with applyable fixes where a safe correction exists:
//
//   off-brand-color  a fill/stroke/text color with no near brand swatch
//                     (when lockColors is on); fix = snap to nearest swatch.
//   off-brand-font   a text family not in the brand font set (when lockFonts
//                     is on); fix = swap to the role-matched brand family.
//   low-contrast     text whose color fails WCAG AA over its background;
//                     fix = nudge the fill to a passing foreground (@hc/color).
//   logo-misuse      a brand logo image placed recolored / non-uniformly
//                     scaled (distorted) / below its min size / inside its
//                     clear space; fix = restore (manual where ambiguous).
//   spacing          an element overlapping the page margin the kit defines
//                     (a pragmatic heuristic, no auto-fix).
//
// Colors are matched in CIELAB via @hc/color (deltaE, nearestPaletteColor),
// reusing the exact tolerance the slice-A locked-save check uses so the live
// lint and the server gate agree. Pure: no IO, runs on the client and a worker.

import type { Color, DesignFile, Fill, Node, Page } from "@hc/schema";
import { contrastRatio, deltaE, fixToAA, nearestPaletteColor } from "@hc/color";
import type { BrandLintViolation, LintBrandKit, LintFont } from "./types";

/** Max perceptual deltaE for a color to count as "matching" a brand swatch.
 *  Mirrors the slice-A locked-save tolerance so lint and the persist gate agree
 *  (below the just-noticeable-difference threshold). */
export const colorTolerance = 2.0;

/** WCAG AA normal-text contrast minimum. */
const AA_NORMAL = 4.5;

/** A logo scaled to a width:height ratio that differs from its natural 1:1
 *  placement by more than this is flagged as distorted (non-uniform scale). */
const ASPECT_TOLERANCE = 0.02;

type AnyRec = Record<string, unknown>;

// --- color extraction -------------------------------------------------------

function eachFillColor(fill: Fill, out: Color[]): void {
  const f = fill as unknown as {
    type?: string;
    color?: Color;
    stops?: { color: Color }[];
  };
  if (f.type === "solid" && f.color) out.push(f.color);
  else if (Array.isArray(f.stops)) for (const s of f.stops) out.push(s.color);
}

interface ColorUse {
  color: Color;
  nodeId: string;
  pageId: string;
}

function collectNodeColors(node: Node, pageId: string, out: ColorUse[]): void {
  const n = node as unknown as {
    id: string;
    fills?: Fill[];
    stroke?: { fill?: Fill; color?: Color };
    content?: { runs: { style: { fill?: Fill; color?: Color } }[] }[];
  };
  for (const fill of n.fills ?? []) {
    const cs: Color[] = [];
    eachFillColor(fill, cs);
    for (const c of cs) out.push({ color: c, nodeId: n.id, pageId });
  }
  if (n.stroke?.fill) {
    const cs: Color[] = [];
    eachFillColor(n.stroke.fill, cs);
    for (const c of cs) out.push({ color: c, nodeId: n.id, pageId });
  }
  if (n.stroke?.color) out.push({ color: n.stroke.color, nodeId: n.id, pageId });
  for (const para of n.content ?? [])
    for (const run of para.runs) {
      if (run.style.fill) {
        const cs: Color[] = [];
        eachFillColor(run.style.fill, cs);
        for (const c of cs) out.push({ color: c, nodeId: n.id, pageId });
      }
      if (run.style.color) out.push({ color: run.style.color, nodeId: n.id, pageId });
    }
}

interface FontUse {
  family: string;
  nodeId: string;
  pageId: string;
}

function collectNodeFonts(node: Node, pageId: string, out: FontUse[]): void {
  const n = node as unknown as {
    id: string;
    type: string;
    content?: { runs: { style: { fontFamily?: string } }[] }[];
  };
  if (n.type !== "text") return;
  for (const para of n.content ?? [])
    for (const run of para.runs) {
      const fam = run.style.fontFamily;
      if (fam) out.push({ family: fam, nodeId: n.id, pageId });
    }
}

function childrenOfNode(node: Node): Node[] {
  const rec = node as unknown as AnyRec;
  const kids = rec.children;
  const out: Node[] = Array.isArray(kids) ? [...(kids as Node[])] : [];
  // Vector composites also nest content (mirrors @hc/templates' findNode).
  if (node.type === "mask" && rec.child) out.push(rec.child as Node);
  if (node.type === "boolean" && Array.isArray(rec.operands))
    out.push(...(rec.operands as Node[]));
  return out;
}

function walk(
  nodes: Node[],
  pageId: string,
  visit: (n: Node, pageId: string) => void,
): void {
  for (const node of nodes) {
    visit(node, pageId);
    const kids = childrenOfNode(node);
    if (kids.length) walk(kids, pageId, visit);
  }
}

// --- helpers ----------------------------------------------------------------

/** Hex for a color (alpha dropped), for messages + fix `from`. */
function hexOf(c: Color): string {
  const ch = (v: number) =>
    Math.round(Math.max(0, Math.min(1, v)) * 255)
      .toString(16)
      .padStart(2, "0");
  return `#${ch(c.srgb.r)}${ch(c.srgb.g)}${ch(c.srgb.b)}`;
}

/** Every approved color in the kit, flattened. */
export function kitColors(kit: LintBrandKit): Color[] {
  const out: Color[] = [];
  for (const pal of kit.palettes) for (const sw of pal.colors) out.push(sw.value);
  return out;
}

/** Lowercased approved font families (for case-insensitive matching). */
export function kitFontFamilies(kit: LintBrandKit): Set<string> {
  return new Set(kit.fonts.map((f) => f.fontFamily.trim().toLowerCase()));
}

/** The brand font to swap an off-brand family to: prefer body, then the first. */
function brandSwapFont(kit: LintBrandKit): LintFont | null {
  return kit.fonts.find((f) => f.role === "body") ?? kit.fonts[0] ?? null;
}

/** The first solid background color of a page, if any (for contrast checks). */
function pageBackgroundColor(page: Page): Color | null {
  if (!page.background) return null;
  const cs: Color[] = [];
  eachFillColor(page.background, cs);
  return cs[0] ?? null;
}

/** The nearest enclosing background color for a text node: its own solid fill is
 *  not the background, so we use the page background (a pragmatic, dependency-
 *  free choice; per-node z-stacking is out of scope for the lint heuristic). */
function backgroundFor(page: Page): Color {
  return pageBackgroundColor(page) ?? { srgb: { r: 1, g: 1, b: 1, a: 1 } };
}

// --- the linter -------------------------------------------------------------

/**
 * Lint a design against a brand kit (FR-7). Returns every violation found, each
 * with a stable id and (where safe) an applyable fix. An on-brand design returns
 * `[]`. Color/font checks only run when the matching lock is on AND the kit
 * actually defines a constraint (an empty palette/font set cannot constrain, so
 * it is treated as "no rule" rather than rejecting everything). Contrast, logo,
 * and spacing checks always run when the kit is present (they are advisory).
 */
export function lintDesign(file: DesignFile, kit: LintBrandKit): BrandLintViolation[] {
  const violations: BrandLintViolation[] = [];
  if (kit.controls.lintPolicy === "off") return violations;

  const palette = kitColors(kit);
  const families = kitFontFamilies(kit);
  const swapFont = brandSwapFont(kit);
  const logoAssetIds = new Set(
    kit.logos.flatMap((l) => [
      l.assetId,
      l.variants?.dark,
      l.variants?.light,
    ]).filter((x): x is string => !!x),
  );
  const logoById = new Map(kit.logos.map((l) => [l.assetId, l]));

  // --- colors + fonts (FR-7 off-brand) --------------------------------------
  if (kit.controls.lockColors && palette.length > 0) {
    const colors: ColorUse[] = [];
    for (const page of file.pages) {
      const bg = pageBackgroundColor(page);
      if (bg) colors.push({ color: bg, nodeId: page.id, pageId: page.id });
      walk(page.children, page.id, (n, pid) => collectNodeColors(n, pid, colors));
    }
    const seen = new Set<string>();
    for (const { color, nodeId, pageId } of colors) {
      const near = nearestPaletteColor(color, palette);
      if (near && near.distance <= colorTolerance) continue;
      const hex = hexOf(color);
      const key = `color:${hex}@${nodeId}`;
      if (seen.has(key)) continue;
      seen.add(key);
      violations.push({
        id: key,
        kind: "off-brand-color",
        severity: "warn",
        nodeId,
        pageId,
        message: `Color ${hex} is not in the brand palette.`,
        fix: near ? { kind: "snap_color", from: hex, to: near.color } : undefined,
      });
    }
  }

  if (kit.controls.lockFonts && families.size > 0) {
    const fonts: FontUse[] = [];
    for (const page of file.pages)
      walk(page.children, page.id, (n, pid) => collectNodeFonts(n, pid, fonts));
    const seen = new Set<string>();
    for (const { family, nodeId, pageId } of fonts) {
      if (families.has(family.trim().toLowerCase())) continue;
      const key = `font:${family}@${nodeId}`;
      if (seen.has(key)) continue;
      seen.add(key);
      violations.push({
        id: key,
        kind: "off-brand-font",
        severity: "warn",
        nodeId,
        pageId,
        message: `Font "${family}" is not an approved brand font.`,
        fix:
          swapFont && swapFont.fontFamily.trim().toLowerCase() !== family.trim().toLowerCase()
            ? { kind: "swap_font", from: family, to: swapFont.fontFamily }
            : undefined,
      });
    }
  }

  // --- low contrast (FR-7) --------------------------------------------------
  for (const page of file.pages) {
    const bg = backgroundFor(page);
    walk(page.children, page.id, (node, pid) => {
      if (node.type !== "text") return;
      const n = node as unknown as {
        id: string;
        content?: { runs: { style: { fill?: Fill; color?: Color } }[] }[];
      };
      // Use the first text color found on the node (the dominant run color).
      let fg: Color | null = null;
      for (const para of n.content ?? []) {
        for (const run of para.runs) {
          if (run.style.color) { fg = run.style.color; break; }
          if (run.style.fill) {
            const cs: Color[] = [];
            eachFillColor(run.style.fill, cs);
            if (cs[0]) { fg = cs[0]; break; }
          }
        }
        if (fg) break;
      }
      if (!fg) return;
      const ratio = contrastRatio(fg, bg);
      if (ratio >= AA_NORMAL) return;
      violations.push({
        id: `contrast:${n.id}`,
        kind: "low-contrast",
        severity: "warn",
        nodeId: n.id,
        pageId: pid,
        message: `Text contrast ${ratio.toFixed(1)}:1 is below WCAG AA (4.5:1).`,
        fix: { kind: "fix_contrast", color: fixToAA(fg, bg, AA_NORMAL) },
      });
    });
  }

  // --- logo misuse (FR-7) ---------------------------------------------------
  if (kit.logos.length > 0) {
    for (const page of file.pages) {
      walk(page.children, page.id, (node, pid) => {
        if (node.type !== "image") return;
        const img = node as unknown as {
          id: string;
          source?: { assetId?: string; src?: string };
          size?: { width: number; height: number };
          transform?: { scaleX?: number; scaleY?: number };
          fills?: Fill[];
          opacity?: number;
        };
        const assetId = img.source?.assetId;
        if (!assetId || !logoAssetIds.has(assetId)) return;
        const logo = logoById.get(assetId);

        // Distortion: non-uniform scale (the image is squashed/stretched).
        const sx = Math.abs(img.transform?.scaleX ?? 1);
        const sy = Math.abs(img.transform?.scaleY ?? 1);
        if (sx > 0 && sy > 0 && Math.abs(sx - sy) / Math.max(sx, sy) > ASPECT_TOLERANCE) {
          violations.push({
            id: `logo-aspect:${img.id}`,
            kind: "logo-misuse",
            severity: "error",
            nodeId: img.id,
            pageId: pid,
            message: "Logo is distorted (non-uniform scale). Scale it uniformly.",
            fix: { kind: "restore_logo", reason: "aspect" },
          });
        }

        // Recolor: a logo image carrying a solid color overlay/tint fill that is
        // not an approved variant is treated as an unapproved recolor.
        if ((img.fills ?? []).some((f) => (f as unknown as AnyRec).type === "solid")) {
          violations.push({
            id: `logo-recolor:${img.id}`,
            kind: "logo-misuse",
            severity: "error",
            nodeId: img.id,
            pageId: pid,
            message: "Logo appears recolored. Use an approved logo variant instead.",
          });
        }

        // Min size: rendered width below the kit's minimum.
        const w = (img.size?.width ?? 0) * sx;
        if (logo?.minSizePx && w > 0 && w < logo.minSizePx) {
          violations.push({
            id: `logo-size:${img.id}`,
            kind: "logo-misuse",
            severity: "warn",
            nodeId: img.id,
            pageId: pid,
            message: `Logo is below its minimum size (${logo.minSizePx}px).`,
            fix: { kind: "restore_logo", reason: "min_size" },
          });
        }
      });
    }
  }

  // --- spacing (FR-7, pragmatic) --------------------------------------------
  // A kit may define a page margin (in px) via controls or a logo clear-space
  // ratio; here we flag elements that overlap the page margin band. The margin
  // is read from the design meta (meta.brandMargin) when present, so this stays
  // a lightweight, dependency-free heuristic that an admin can opt into.
  const margin = readMargin(file);
  if (margin > 0) {
    for (const page of file.pages) {
      const w = page.width;
      const h = page.height;
      walk(page.children, page.id, (node, pid) => {
        const t = (node as unknown as AnyRec).transform as { x?: number; y?: number } | undefined;
        const s = (node as unknown as AnyRec).size as { width?: number; height?: number } | undefined;
        if (!t || !s) return;
        const x = t.x ?? 0;
        const y = t.y ?? 0;
        const nw = s.width ?? 0;
        const nh = s.height ?? 0;
        const overlapsMargin =
          x < margin || y < margin || x + nw > w - margin || y + nh > h - margin;
        if (overlapsMargin) {
          violations.push({
            id: `spacing:${node.id}`,
            kind: "spacing",
            severity: "info",
            nodeId: node.id,
            pageId: pid,
            message: `Element extends into the ${margin}px brand margin.`,
          });
        }
      });
    }
  }

  return violations;
}

/** The opt-in page margin (px) for the spacing check, from `meta.brandMargin`. */
function readMargin(file: DesignFile): number {
  const m = (file.meta as AnyRec)?.brandMargin;
  return typeof m === "number" && m > 0 ? m : 0;
}

export { hexOf };
